src/service/resources-impl.js

```diff
@@ -281,94 +287,144 @@ export class ResourcesImpl {
 
   /** @override */
   isIntersectionExperimentOn() {
-    return !!this.intersectionObserver_;
+    return !!this.oneViewportObserver_;
   }
 
   /**
    * @param {!Array<!IntersectionObserverEntry>} entries
-   * @param {!IntersectionObserver} unusedObserver
    * @private
    */
-  intersects_(entries, unusedObserver) {
-    dev().fine(TAG_, 'intersect', entries);
+  intersectsOneViewport_(entries) {
+    // TODO(willchou): Remove assert once #27167 is fixed.
+    devAssert(this.prerenderSize_ == 1);
+
+    // Store the 1vp rect for later.
+    if (!this.oneViewportRect_) {
+      this.oneViewportRect_ =
+        /** @type {!ClientRect} */ (entries[0].rootBounds);
+    }
 
-    const toUnload = [];
+    const whenBuilt = this.intersects_(entries, /* viewports */ 1);
+
+    // To prioritize in-viewport elements, delay handling of 2vp/3vp until
+    // all 1vp elements are built (and we're not in prerender, of course).
+    Promise.all([whenBuilt, this.ampdoc.whenFirstVisible()]).then(() => {
+      if (this.threeViewportsObserver_) {
+        return;
+      }
+      this.threeViewportsObserver_ = new IntersectionObserver(
+        e => this.intersectsThreeViewports_(e),
+        {
+          root: this.oneViewportObserver_.root,
+          rootMargin: '200% 25%',
+        }
+      );
+      const elements = this.resources_.map(r => r.element);
```

Comment:
Is there value in doing a `map` then a `forEach`? Or does this indirection get removed by Closure Compiler?

```javascript
this.resources.forEach(r => (
  e.element.whenUpgraded().then(() => 
    this.threeViewportsObserver_.observe(r.element)
  )
);
```

Reply:
This used to be more complex -- I'll simplify. 👍